frameworks/CMSIS/LPC1768/lib/usb/mscuser.h

```diff
@@ -49,21 +49,56 @@ enum class Sense_KEY : uint8_t {
   DATA_PROTECT
 };
 
-enum class Sense_ASC : uint8_t {
+enum struct Sense_ASC : uint8_t {
+  NONE = 0x0,
+  LOGICAL_UNIT_NOT_READY = 0x04,
   CANNOT_READ_MEDIUM = 0x30,
-  MEDIUM_NOT_PRESENT = 0x3A
+  MEDIUM_NOT_PRESENT = 0x3A,
+  MEDIA_CHANGED = 0x28
 };
 
-enum class Sense_ASCQ : uint8_t {
-  // CANNOT_READ_MEDIUM
+enum struct Sense_ASCQ : uint8_t {
+  // ASC: LOGICAL_UNIT_NOT_READY
+  CAUSE_NOT_REPORTABLE = 0x00,
+  UNIT_IS_IN_PROCESS_OF_BECOMING_READY,
+  INITIALIZING_COMMAND_REQUIRED,
+  MANUAL_INTERVENTION_REQUIRED,
+  FORMAT_IN_PROGRESS,
+  DEVICE_IS_BUSY = 0xFF,
+  // ASC: CANNOT_READ_MEDIUM
   UNKNOWN_FORMAT = 0x01,
-  // MEDIUM_NOT_PRESENT
+  // ASC: MEDIUM_NOT_PRESENT
   REASON_UNKNOWN = 0x00,
   TRAY_CLOSED,
   TRAY_OPEN,
   LOADABLE,
   AUXILIARY_MEMORY_ACCESSIBLE
 };
 
+struct Sense {
+  Sense() {
+    reset();
+  }
+
+  void set(Sense_KEY key_val, Sense_ASC asc_val = Sense_ASC::NONE, Sense_ASCQ ascq_val = Sense_ASCQ::REASON_UNKNOWN) {
+    key = key_val;
+    asc = asc_val;
+    ascq = ascq_val;
+  }
+
+  void reset() {
+    key = Sense_KEY::NO_SENSE;
+    asc = Sense_ASC::NONE;
+    ascq = Sense_ASCQ::REASON_UNKNOWN;
+  }
+
+  bool has_sense() {
+    return key != Sense_KEY::NO_SENSE;
+  }
+
+  Sense_KEY key;
+  Sense_ASC asc;
+  Sense_ASCQ ascq;
+};
```

Comment:
If there will only ever be one `Sense` object, this class can be made all-static to use function-like calling semantics, which are slightly smaller and faster. I realize this is far less important on 32-bit platforms, but it would be consistent. That said, this class does look like more than one instance could be used.

Reply:
You're right this class could be static its still very much a work in progress, I'm still hoping I can find a nice compatible, fully functional usb stack to replace all this with, I'm not familiar enough with the intricacies (although a lot more now than I was) and workarounds for OS differences that are needed. (if only Windows used the Lock/Unlock functionality built into the damn protocol)